driver-async/src/main/com/mongodb/async/client/ClientSessionImpl.java

```diff
@@ -55,9 +55,16 @@ public boolean hasActiveTransaction() {
 
     @Override
     public boolean notifyMessageSent() {
-        boolean firstMessage = !messageSent;
-        messageSent = true;
-        return firstMessage;
+        if (hasActiveTransaction()) {
```

Comment:
Now this method is called whether in or out of a transaction, so have to handle both cases.